src/ai/defaultai.cc

```diff
@@ -5856,7 +5830,22 @@ BuildingNecessity DefaultAI::check_building_necessity(BuildingObserver& bo,
 			} else {
 				bo.cnt_target = 1 + static_cast<int32_t>(mines_.size() + productionsites.size()) / 30;
 			}
+			// Determine whether we need more buildings due to low stocklevel
+			const uint32_t current_stocklevel = (get_stocklevel(bo, gametime));
+			const uint8_t stocklevel_threshhold =
+			   std::abs(management_data.get_military_number_at(179));
 
+			if (current_stocklevel > stocklevel_threshhold &&
+			    persistent_data->remaining_basic_buildings.count(bo.id) == 0) {
+				bo.cnt_target -= 1;
+			}
```

Comment:
But you might get below basic_amount, cant you?

Reply:
```
if (current_stocklevel > stocklevel_threshhold &&
    bo.cnt_target > bo.basic_amount) {
	bo.cnt_target -= 1;
}
```

Reply:
good point will add the check for basic economy established as a safeguard.

Reply:
I meant I would not involve  persistent_data->remaining_basic_buildings here at all

Reply:
Now I am confused.  I overlooked that I used persistent_data->remaining_basic_buildings in my first answer. This is directly related to whether there is still a building needed to establish basic economy. After having reached basic economy basic_amount does not matter anymore afaik.  So I tend to change nothing here

Reply:
> After having reached basic economy basic_amount does not matter anymore afaik. 

This is kind of undefined. I would vote never go below basic economy. Anyway the code is hard to read, so perhaps:

```
if (current_stocklevel > stocklevel_threshhold &&
		basic_economy_achieved && bo.cnt_target > 0) {
		bo.cnt_target -= 1;
	}
```
This is what you mean?

Reply:
Basic economy buildings counts are so low anyway - I see no benefit dismantling them in 95% of cases

Reply:
this code isn't about dismantling at all it is only for construction

Reply:
There is other code that handles possible dismantling. I can differ, but generally logic should be never dismantle under cnt_target, of course the implementation in check_productionsites() can differ for various types of sites
So I suggest that cnt_target should never get below basic_amount

Reply:
or we can have two cnt_targets: 

- cnt_min_taret - for dismantling
- cnt_max_target - for building new ones

Reply:
I carefully checked the dismantling code and the target is only one safeguard of many others.  Basically dismantling is based on high stocklevel and low performance of a specific site. So if the site is running at 0% and is only protected by cnt_taget as last safeguard it is questionable if this would really be a good idea.  
So I am really 100% sure that this is safe and it gives us just the little push we need.